ecosystem/typescript/sdk/src/transaction_builder/builder_utils.ts

```diff
@@ -0,0 +1,294 @@
+  if (c === ":") {
+    if (tagStr.slice(pos, pos + 2) === "::") {
+      return [["COLON", "::"], 2];
+    }
+    bail("Unrecognized token.");
+  } else if (c === "<") {
+    return [["LT", "<"], 1];
+  } else if (c === ">") {
+    return [["GT", ">"], 1];
+  } else if (c === ",") {
+    return [["COMMA", ","], 1];
+  } else if (isWhiteSpace(c)) {
+    let res = "";
+    for (let i = pos; i < tagStr.length; i += 1) {
+      const char = tagStr[i];
+      if (isWhiteSpace(char)) {
+        res = `${res}${char}`;
+      } else {
+        break;
+      }
+    }
+    return [["SPACE", res], res.length];
+  } else if (isValidAlphabetic(c)) {
+    let res = "";
+    for (let i = pos; i < tagStr.length; i += 1) {
+      const char = tagStr[i];
+      if (isValidAlphabetic(char)) {
+        res = `${res}${char}`;
+      } else {
+        break;
+      }
+    }
+    return [["IDENT", res], res.length];
+  }
+  throw new Error("Unrecognized token.");
+}
+
+function tokenize(tagStr: string): Token[] {
+  let pos = 0;
+  const tokens = [];
+  while (pos < tagStr.length) {
+    const [token, size] = nextToken(tagStr, pos);
+    if (token[0] !== "SPACE") {
+      tokens.push(token);
+    }
+    pos += size;
+  }
+  return tokens;
+}
+
+/**
+ * Parser to parse a type tag string
+ */
+export class TypeTagParser {
+  private readonly tokens: Token[];
+
+  constructor(tagStr: string) {
+    this.tokens = tokenize(tagStr);
+  }
+
+  private consume(targetToken: string) {
+    const token = this.tokens.shift();
+    if (!token || token[1] !== targetToken) {
+      bail("Invalid type tag.");
+    }
+  }
+
+  private parseCommaList(endToken: TokenValue, allowTraillingComma: boolean): TypeTag[] {
+    const res: TypeTag[] = [];
+    if (this.tokens.length <= 0) {
+      bail("Invalid type tag.");
+    }
+
+    while (this.tokens[0][1] !== endToken) {
+      res.push(this.parseTypeTag());
+
+      if (this.tokens.length > 0 && this.tokens[0][1] === endToken) {
+        break;
+      }
+
+      this.consume(",");
+      if (this.tokens.length > 0 && this.tokens[0][1] === endToken && allowTraillingComma) {
+        break;
+      }
+
+      if (this.tokens.length <= 0) {
+        bail("Invalid type tag.");
+      }
+    }
+    return res;
+  }
+
+  parseTypeTag(): TypeTag {
+    if (this.tokens.length === 0) {
+      bail("Invalid type tag.");
+    }
+
+    // Pop left most element out
+    const [tokenTy, tokenVal] = this.tokens.shift();
+
+    if (tokenVal === "u8") {
+      return new TypeTagU8();
+    }
+    if (tokenVal === "u64") {
+      return new TypeTagU64();
+    }
+    if (tokenVal === "u128") {
+      return new TypeTagU128();
+    }
+    if (tokenVal === "bool") {
+      return new TypeTagBool();
+    }
+    if (tokenVal === "address") {
+      return new TypeTagAddress();
+    }
+    if (tokenVal === "vector") {
+      this.consume("<");
+      const res = this.parseTypeTag();
+      this.consume(">");
+      return new TypeTagVector(res);
```

Comment:
do we support nested vector?  nested vector of primitive types are all valid

Reply:
Yes. It supports nested vectors.

Reply:
For a module script function, yes, we do support embedded vectors. But for a script, the arguments can only be a vector of u8s. See my other comment.